### PR TITLE
Added ability to access and send message headers in requests

### PR DESCRIPTION
**Pull request by mhelleborg**
Alternative to solve #1024, and support richer use cases for headers. Breaking change for future-tasks, as they no longer unpack message envelopes.

### CODE REVIEW

**Comment by mhelleborg** [comment]
We could possibly drop the MessageHeader parameter to `RequestWithHeadersAsync`, since it could just take a MessageEnvelope instead.  It would be less obvious to newcomers, but then again, they likely do not care about headers :)

File: src/Proto.Actor/Context/ActorContext.cs [hunk 1]
```diff
@@ -65,6 +65,8 @@ public void Stash()
             if (_messageOrEnvelope is not null) EnsureExtras().Stash.Push(_messageOrEnvelope);
         }
 
+        public void Respond(object message, MessageHeader header) => Respond(new MessageEnvelope(message, null, header));
```
Comment (rogeralsing) [inline src/Proto.Actor/Context/ActorContext.cs]: 👍🏻

File: src/Proto.Actor/Context/IContext.cs [hunk 2]
```diff
@@ -33,6 +33,13 @@ public interface IContext : ISenderContext, IReceiverContext, ISpawnerContext, I
         /// </summary>
         /// <param name="message">The message to send</param>
         void Respond(object message);
+        
+        /// <summary>
+        ///     Sends a response to the current Sender, including message header
+        /// </summary>
+        /// <param name="message">The message to send</param>
+        /// <param name="header"></param>
+        void Respond(object message, MessageHeader header) => Respond(new MessageEnvelope(message, null, header));
```
Comment (rogeralsing) [inline src/Proto.Actor/Context/IContext.cs:42]: 👍🏻

File: src/Proto.Actor/Context/ISenderContext.cs [hunk 3]
```diff
@@ -102,23 +102,65 @@ public static class SenderContextExtensions
...
+        )
+        {
+            using var future = self.GetFuture();
+            var messageEnvelope = new MessageEnvelope(message, future.Pid, headers);
+            self.Send(target, messageEnvelope);
+            var result = await future.GetTask(cancellationToken);
+
+            var messageResult = MessageEnvelope.UnwrapMessage(result);
+
+            switch (messageResult)
```
Comment (rogeralsing) [inline src/Proto.Actor/Context/ISenderContext.cs:129]: We have this switch a bit all over the place afaik. can it be moved to a shared method?

File: src/Proto.Actor/Context/ISenderContext.cs [hunk 4]
```diff
@@ -102,23 +102,65 @@ public static class SenderContextExtensions
...
+                        $"Unexpected message. Was type {messageResult.GetType()} but expected {typeof(T)}"
+                    );
+            }
+        }
+
+        internal static async Task<T> RequestAsync<T>(this ISenderContext self, PID target, object message, CancellationToken cancellationToken)
         {
             using var future = self.GetFuture();
-            var messageEnvelope = new MessageEnvelope(message, future.Pid);
+            var messageEnvelope = message is MessageEnvelope envelope ? envelope.WithSender(future.Pid) : new MessageEnvelope(message, future.Pid);
```
Comment (rogeralsing) [inline src/Proto.Actor/Context/ISenderContext.cs:144]: Is this to allow us to explicitly pass MessageEnvelopes as a message? (I assume yes, but just to be 100% :) )

File: src/Proto.Actor/Context/ISenderContext.cs [hunk 5]
```diff
@@ -102,23 +102,65 @@ public static class SenderContextExtensions
...
             self.Send(target, messageEnvelope);
             var result = await future.GetTask(cancellationToken);
 
-            switch (result)
+            var messageResult = MessageEnvelope.UnwrapMessage(result);
+
+            switch (messageResult)
```
Comment (rogeralsing) [inline src/Proto.Actor/Context/ISenderContext.cs:150]: See above comment on repeated code

File: src/Proto.Actor/Context/ISenderContext.cs [hunk 6]
```diff
@@ -102,23 +102,65 @@ public static class SenderContextExtensions
...
+
+            var messageResult = MessageEnvelope.UnwrapMessage(result);
+
+            switch (messageResult)
+            {
+                case DeadLetterResponse:
+                    throw new DeadLetterException(target);
+                case null:
+                case T:
+                    return ((T) messageResult!, MessageEnvelope.UnwrapHeader(result));
```
Comment (rogeralsing) [inline src/Proto.Actor/Context/ISenderContext.cs:133]: My bad regarding repeated code, this bit is ofc different.

**Comment by rogeralsing** [comment]
>We could possibly drop the MessageHeader parameter to RequestWithHeadersAsync

I like the consistency, if we get message + headers back from the task, it makes sense to pass in message + headers into the request

File: src/Proto.Actor/Context/ISenderContext.cs (hunk 4 above)
Comment (mhelleborg) [reply src/Proto.Actor/Context/ISenderContext.cs:144]: It is indeed. Added a test for it as well

File: src/Proto.Actor/Messages/MessageEnvelope.cs [hunk 7]
```diff
@@ -35,6 +38,16 @@ public MessageEnvelope(object message, PID? sender, MessageHeader? header = null
         public MessageEnvelope WithHeader(MessageHeader header) =>
             this with {Header = header};
 
+        public MessageEnvelope MergeHeader(MessageHeader header)
```
Comment (rogeralsing) [inline src/Proto.Actor/Messages/MessageEnvelope.cs:41]: 👍🏻

**Pull request merged**